proposals/interface-types/Explainer.md

```diff
@@ -0,0 +1,658 @@
+text format (as recommended by the [Custom Annotation] proposal). This `func`
+statement defines two things:
+* an export `greeting` in the [adapted module](#Overview) whose result is the
+  `string` interface type
+* an [adapter function](#Overview) that implements `greeting`
+
+While the set of *instructions* in the body of the adapter function is
+completely different from the set of [core wasm instructions][Instructions],
+the overall *structure* of the adapter module is the same as a core WebAssembly
+function. In particular:
+* The adapter function body is a (possibly nested) sequence of instructions.
+* Instructions are defined to pop operands and push results onto the stack.
+* Validation checks that the types of instructions' signatures line up.
+* Instructions can either be written as a linear sequence (as shown above)
+  or, equivalently, as [Folded S-expressions].
+
+The specific adapter function shown above uses two adapter instructions:
+* The `call-export` instruction calls the `greeting_` export of the core wasm
+  module which leaves two `i32` values on the stack.
+* The `memory-to-string` instruction pops the two `i32` values, reads and
+  decodes UTF-8 bytes from `mem`, and pushes the resulting `string` value.
+
+Having added this adapter function, the `.wat` file now defines an [adapted module](#Overview)
+with no imports and whose only export is the (adapted) `greeting` export.
+Importantly, the `greeting_` and `mem` core-module exports are not exported
+from the adapted module: they are fully encapsulated by the adapted module.
+While it is possible for an adapted module to re-export core-module exports,
+this must be explicitly specified with an `@interface` statement.
+
+Another note is that, since `greeting_` and `greeting` are in separate export
+namespaces (of the core and adapted modules, resp.), they could have the same
+name. The trailing underscore is only added for clarity in this walkthrough.
+
+### Export returning string (dynamically allocated)
+
+The previous section's example has the simplifying property that the string is
+statically allocated. What happens if `greeting_` wants to `malloc` the
+resulting string? For example:
+
+```wasm
+(module
+  (memory (export "mem") 1)
+  (func $malloc (param i32) (result i32) ...)
+  (func (export "free") (param i32) ...)
+  (func (export "greeting_") (result i32 i32)
+    ;; compute allocation size
+    call $malloc
+    ;; initialize malloc'd memory with string
+    ;; return pointer and length
+    ;; caller must call "free" when done the string
+  )
+)
+```
+
+This situation would naturally arise in, e.g., a C++ function that returns a
+`std::string` or `std::unique_ptr<char[]>`.
+
+What we need is to be able to call `free` right after the bytes are
+read from `memory-to-string`. Thus, `memory-to-string` takes an
+optional exported function name that it calls after it has read the string:
```

Comment:
Having an optional export param seems less elegant than `call-export "free"` on the pointer, but that does require stack modification (`dup` et al) or locals in these functions, both of which seem more complex than is worth it just for this. Might be worth keeping an eye open for where it might make sense in the future.

Reply:
I totally know what you mean and indeed I originally rewrote this example to do the stack modifications (with `pick`, `swap`, and `drop`), but that all ends up being a bit ugly, so I considered adding [`let`](https://github.com/WebAssembly/function-references/blob/master/proposals/function-references/Overview.md#local-bindings), but that's complicated too.  Then I realized that all of this needs to be made exception safe and how there's probably some elegant, succinct RAII-like instruction that handles all these cases, so I put in the exception-safety note.  But I think I should be more explicit about this connection, so I updated that note. lmkwyt!

Reply:
I think there may be a bigger issue here. The call to the exported function is 'in the middle' of a call to the corresponding import. The 'free' will happen before the return to the caller; which means that the ...-to-string call must copy the data because the next thing that will happen is a free of the string data -- before returning to the caller.
What is really needed, if feasible, is that the 'free' happens after the caller has finished with the string. But this seems problematic.
Perhaps a RUST-style ownership model may work ...

Reply:
Ah, hah, I see your point: according to a plain operational semantics, we'd need to call "free" in the callee before we can call "malloc" in the caller (to get a `memcpy` destination).  I think one option is to tie the free call to the `string` value, so that "free" is called only after the `string` goes out of scope, which is I think what you mean.  But then we still have to worry about exception safety between the memory being allocated (by the final wasm callee) and `memory-to-string`.  I'm imagining there is some instruction that you can call immediately after you get a freshly-allocated `i32` pointer and it calls `free` at the end of the *entire adapter-pair* call.  How about I change the note to indicate that we need to solve *both* these problems?

Reply:
Pushed a new version of the note; PTAL.

Reply:
Re: ownership, see also: https://github.com/WebAssembly/interface-types/issues/42#issuecomment-520410978